src/MOI_wrapper.jl

```diff
@@ -50,22 +59,40 @@ mutable struct Optimizer <: MOI.AbstractOptimizer
     end
 end
 
+"Release references and free memory."
+function free_scip(o::Optimizer)
+    # Avoid double-free (SCIP will set the pointers to NULL).
+    if o.inner.scip[] != C_NULL
+        for c in values(o.inner.conss)
+            @SCIP_CALL SCIPreleaseCons(o.inner, c)
+        end
+        for v in values(o.inner.vars)
+            @SCIP_CALL SCIPreleaseVar(o.inner, v)
+        end
+        # only o.inner.scip is GC-protected during ccall!
+        GC.@preserve o begin
```

Comment:
I'm not sure whether this is correct. Maybe the `GC.@preserve` is not needed at all, because this function already has `o` as an argument. It could also be that `o.inner` should be passed to the `GC.@preserve`. I forgot the specific need for this macro call.

Reply:
You should implement `cconvert` for the type instead of using `@preserve` https://jump.dev/MathOptInterface.jl/stable/manual/implementing/#Optimizer-objects-for-C-solvers

Reply:
Thanks for the clarification. There's already an implementation of [unsafe_convert](https://github.com/scipopt/SCIP.jl/blob/0d0321d31ae177a71228c8055274ebbd047b40ca/src/scip_data.jl#L35) for `SCIPData`.